src/python/tests/test_local_functions.py

```diff
@@ -0,0 +1,258 @@
+# software.
+########################################################################
+
+import itertools
+import json
+import numpy as np
+import os
+
+from numpy.testing import assert_approx_equal
+from unittest import TestCase
+
+from pygrackle import \
+    FluidContainer, \
+    chemistry_data, \
+    setup_fluid_container
+
+from pygrackle.utilities.physical_constants import \
+    cm_per_mpc, \
+    mass_hydrogen_cgs, \
+    sec_per_Myr
+from pygrackle.utilities.testing import \
+    ensure_dir, \
+    generate_local_function_results, \
+    grackle_data_dir, \
+    local_function_test_format_version, \
+    test_answers_dir
+
+_meta_data = {"format_version": local_function_test_format_version}
+
+ensure_dir(test_answers_dir)
+
+parameter_grid = {
+    "use_grackle": (1,),
+    "primordial_chemistry": (0, 1, 2, 3),
+    "dust_chemistry": (0, 1),
+    "metal_cooling": (0, 1),
+    "UVbackground": (0, 1),
+    "cmb_temperature_floor": (0, 1),
+    "grackle_data_file": ("CloudyData_UVB=HM2012.h5",),
+}
+
+exclude_sets = (
+    {"dust_chemistry": 1, "metal_cooling": 0},
+    {"dust_chemistry": 1, "primordial_chemistry": 0},
+    {"dust_chemistry": 1, "primordial_chemistry": 1},
+)
+
+grackle_functions = (
+    "cooling_time",
+    "dust_temperature",
+    "gamma",
+    "pressure",
+    "temperature"
+)
+
+function_requirements = {
+    "dust_temperature": {"dust_chemistry": 1},
+}
+
+base_units = {
+    "comoving_coordinates": 1,
+    "a_units": 0.1,
+    "density_units": mass_hydrogen_cgs,
+    "length_units": cm_per_mpc,
+    "time_units": sec_per_Myr
+}
+
+random_inputs = {
+    "density":     {"min": -27, "max":  -20, "log": True},
+    "temperature": {"min":   1, "max":    8, "log": True},
+    "metallicity": {"min":  -6, "max": -1.5, "log": True},
+    "redshift":    {"min":   0, "max":    6, "log": False},
+}
+
+class LocalFunctionsTest(TestCase):
+    """
+    Tests for the local functions.
+    """
+
+    n_input_sets = 10
+    seed = 21062024
+    digits = 12
+    test_file = os.path.join(test_answers_dir, "local_function_tests.json")
+
+    def setUp(self):
+        """
+        Setup local function tests.
+
+        If we are generating results, then create a list of parameter values
+        from the parameter_grid and exclude_sets structures defined above.
+        Also, use the base_units dict to create units.
+
+        If we are not generating results, then read everything in from the
+        json test file.
+        """
+
+        if generate_local_function_results:
+            my_sets = self.generate_parameter_sets()
+            self.test_sets = [{"parameters": my_set, "units": base_units}
+                              for my_set in my_sets]
+        else:
+            if not os.path.exists(self.test_file):
+                self.skipTest(f"Test file not found: {self.test_file}.")
+            with open(self.test_file, mode="r") as f:
+                self.test_sets = json.load(f)
```

Comment:
@mabruzzo, here is where we either generate the test inputs internally or load them from the json file.